Major rewrite, to fullfill #4 and #6

Rocco83 opened the pull request:
Hi,

i am sure that this can be coded better.
but seems working (still noop, to start sharing).

comments are welcome.

Rocco83 committed:
Fixing issue #4 and #6

Rocco83 committed:
Fixed https instead of http, introduced by 59dd382

evgeni commented:
hey, could you please rebase?

I promise I'll review it then :)

Rocco83 committed:
Aligned with commit 106ef34

Rocco83 committed:
Fixed tab with space

Rocco83 committed:
Merge branch 'master' of https://github.com/RedHatSatellite/katello-attach-subscription

Conflicts:
	katello-attach-subscription
	katello-attach-subscription.yaml

Rocco83 commented:
I have managed the rebase.
Code still not tested, and also the logic has to be reviewed since the code changed heavily on both side.

evgeni commented on katello-attach-subscription:
```diff
@@ -15,9 +15,12 @@
 require 'optparse'
 require 'yaml'
 require 'apipie-bindings'
+#require 'pry'
```
what are those new requires for? I'd prefer not to have any gems loaded that are not shipped in RHEL or the satellite6-tools repo.

evgeni commented on katello-attach-subscription:
```diff
@@ -84,91 +95,169 @@ end
+        next
       end
 
+      #p sys['type']
```
iiih, commented code ;)

evgeni commented on katello-attach-subscription:
```diff
@@ -84,91 +95,169 @@ end
       end
+      # extract the name of the host to be registered
       hostnameregex = Regexp.new(sub['hostname'])
+      STDERR.puts sub['hostname'].inspect
```
we probably should have a --debug switch, but not STDERR.puts

evgeni commented on katello-attach-subscription:
```diff
@@ -84,91 +95,169 @@ end
 
-    if desired_sub
-      has_desired_sub = false
+    #p desired_sub_hash
```
commented code

Rocco83 committed:
Working and tested version

Rocco83 commented:
Hi there, please check and review

Rocco83 committed:
Code cleanup

Rocco83 committed:
Added Gemfile for bundle support

Rocco83 committed:
Documentation and yaml file upgrade
Documentation:
* updated the command line and yaml settings description
* changed yaml example structure
* added more description on what registered-by means
.gitignore
* added katello-attach-subscription.yaml
YAML file
* renamed to -example

Rocco83 commented:
Any news? :-)

Rocco83 committed:
Fixed syntax error in cache file concatenation

Rocco83 committed:
Bug fix and debug improved

The subscription associated to the system has been moved inside the
loop. This is less efficient (reiterate the call), but the current
methodology give 404 error on the second loop.
Added debug in case of subscription not existing
Clarified the debug when a subscription cannot be attached

Rocco83 commented:
Code reviewed and tested.

evgeni commented on .gitignore:
```diff
@@ -0,0 +1,54 @@
+.repl_history
+build/
+*.bridgesupport
+build-iPhoneOS/
```
I don't see how all these entries are relevant for us? dot-gem, okay, our own configs, okay. but the rest? :)

Rocco83 replied:
Taken from here 
https://github.com/github/gitignore/blob/master/Ruby.gitignore
And forgot to add a comment line.

spec, test and temp i think is useful in the future.
We can remove instead iPhone and other stuff, but isn't better to have a reference?

evgeni commented on katello-attach-subscription:
```diff
@@ -77,98 +95,239 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+  # satellite url has to start with https or PUT will fail with http error
+  unless @options[:uri].start_with?('https://')
+    puts " FATAL ERROR: the uri must start with https://"
+    exit fail
```
you want `abort "FATAL ERROR: the uri must start with https://"` instead of puts&exit

also `exit fail` is invalid as `fail` is not defined.

evgeni commented on katello-attach-subscription:
```diff
@@ -77,98 +95,239 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+      systems = YAML.load_file(cachefile)
+    rescue Exception
+      puts " FATAL ERROR: Failed to read cache file. re-run without --use-cache."
+      exit false
```
`abort …`

evgeni commented on katello-attach-subscription:
```diff
@@ -77,98 +95,239 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+        end
+        # check a single subscription in the array
+        req['results'].each do |sub|
+          # move subscription_id key to cp_id, to have a standard API interface
```
don't like this -- but I see your point :(

evgeni commented:
could you remove the commented out code that is not needed anymore please? :-)
I also left a few minor inline comments, but as my lab is broken atm I cannot test the code :(

Rocco83 committed:
Code cleanup

Added reference for gitignore file
Fixed typo exit false
Removed lines commented lines of code

evgeni commented on katello-attach-subscription:
```diff
@@ -77,98 +95,235 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+              puts " ERROR: unable to retrieve subscription #{desired_sub}"
+              puts e.message
+              puts e.backtrace.inspect
+              exit fail
```
`exit 1`

evgeni commented on katello-attach-subscription:
```diff
@@ -77,98 +95,235 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+                puts " ERROR: unable to attach subscription"
+                puts e.message
+                puts e.backtrace.inspect
+                exit fail
```
`exit 1`

Rocco83 committed:
Fixed exit false with exit 1.

evgeni commented:
Mhh. Something is wrong with the code I fear :(

I have a host, which Sat 6.2 reports (on https://sat62.egolov.example.com/api/v2/hosts/16/subscriptions) to have the following subscription data:

```
{
    "error": null,
    "page": 1,
    "per_page": 1,
    "results": [
        {
            "account_number": 5597161,
            "available": 4,
            "consumed": 1,
            "contract_number": 11058188,
            "cores": null,
            "cp_id": "4028fa885723ddeb015723e61f6b023c",
            "end_date": "2017-08-15T03:59:59.000+0000",
            "id": 7,
            "instance_multiplier": 1,
            "multi_entitlement": null,
            "name": "Employee SKU",
            "product_id": "ES0113909",
            "product_name": "Employee SKU",
            "quantity": 5,
            "quantity_consumed": 1,
            "ram": null,
            "sockets": 128,
            "stacking_id": null,
            "start_date": "2016-08-15T04:00:00.000+0000",
            "subscription_id": 6,
            "support_level": "Self-Support",
            "type": "NORMAL",
            "unmapped_guest": false,
            "virt_only": false
        }
    ],
    "search": null,
    "sort": {
        "by": null,
        "order": null
    },
    "subtotal": 1,
    "total": 1
}
```

With the old code, and the following yaml:
```
  - hostname: rhel71.egolov.example.com
    type: Virtual Guest 
    sub: 4028fa885723ddeb015723e61f6b023c
```
I get:
```
rhel71.egolov.example.com (bbb5ff74-5758-4637-ab89-54f01aed439d)
 found 4028fa885723ddeb015723e61f6b023c
```
→ nothing happens.

With your code, and the adopted yaml:
```
  - hostname: rhel71.egolov.example.com
    type: Virtual Guest 
    sub:
      rhel:
        - 4028fa885723ddeb015723e61f6b023c
```
I get:
```
Current system: rhel71.egolov.example.com (bbb5ff74-5758-4637-ab89-54f01aed439d)
 Adding subscription for bbb5ff74-5758-4637-ab89-54f01aed439d
in the desired_sub_hash ["4028fa885723ddeb015723e61f6b023c"]
 removing subscription 6 from system bbb5ff74-5758-4637-ab89-54f01aed439d
 [noop] removed
Subscription for product rhel currently missing. Set for the attach.
 adding 4028fa885723ddeb015723e61f6b023c for rhel (id: 7)
 [noop] added 4028fa885723ddeb015723e61f6b023c for rhel in system bbb5ff74-5758-4637-ab89-54f01aed439d
```

So it seems to remove the sub just to add it again.
I think this is some id vs cp_id vs subscription_id mixup?

evgeni commented:
```
          if (sub.has_key?('subscription_id'))
            sub['cp_id'] = sub.delete('subscription_id')
          end
```
that's the offender here.

Rocco83 committed:
Fixed problem for Katello 3.0 (Satellite 6.2) always removing and adding subs
Less debug in standard output if not with --debug
Removed commented code

Rocco83 committed:
Fixed check on integer for subscription_id

Rocco83 committed:
Fixed typo in subscription_id integer check
